Modules/Loadable/Colors/MRMLDM/vtkMRMLColorLegendDisplayableManager.cxx

```diff
@@ -254,19 +296,34 @@ bool vtkMRMLColorLegendDisplayableManager::vtkInternal::UpdateActor(vtkMRMLColor
     vtkMRMLVolumeDisplayNode* volumeDisplayNode = vtkMRMLVolumeDisplayNode::SafeDownCast(primaryDisplayNode);
     if (volumeDisplayNode)
       {
+      vtkMRMLVolumeNode* volumeNode = vtkMRMLVolumeNode::SafeDownCast(volumeDisplayNode->GetDisplayableNode());
       // Volumes are special case, their visibility can be determined from slice view logics
       if (this->SliceCompositeNode)
         {
         // 2D view
-        visible &= this->IsVolumeVisibleInSliceView(vtkMRMLVolumeNode::SafeDownCast(volumeDisplayNode->GetDisplayableNode()));
+        visible &= this->IsVolumeVisibleInSliceView(volumeNode);
         }
       else
         {
         // 3D view
-        // For now don't show color legends for volumes in 3D views.
-        // In the future, color legend could be shown for volumes that are shown in slice views
+        // Color legend is be shown for volumes that are shown in slice views
         // that are visible in the 3D view.
-        visible = false;
+        std::vector< vtkMRMLNode* > sliceNodes;
+        this->External->GetMRMLScene()->GetNodesByClass("vtkMRMLSliceNode", sliceNodes);
+        for (auto aNode : sliceNodes)
+          {
+          vtkMRMLSliceNode* sliceNode = vtkMRMLSliceNode::SafeDownCast(aNode);
+          if (sliceNode && sliceNode->GetSliceVisible())
+            {
+            vtkMRMLSliceCompositeNode* sliceCompositeNode = vtkMRMLSliceLogic::GetSliceCompositeNode(sliceNode);
+            visible &= this->IsVolumeVisibleInSliceView(sliceCompositeNode, volumeNode);
```

Comment:
This would pollute all 3D viewers with color legends from all slice views.

A little improvement would be to ignore slice view nodes in different view groups as the 3D view node (then, for example in a 4-over-4 view, slice nodes in the top row would not make a color legend show up in the bottom row).

However, it would be better to only show the color legend if the slice is shown in that 3D view (eye icon is enabled in the slice view). It could work like this:

```
            vtkMRMLModelNode* sliceModelNode = sliceLogic->GetSliceModelNode();
            vtkMRMLDisplayNode* sliceModelDisplayNode = (sliceModelNode ? sliceModelNode->GetDisplayNode() : nullptr);
            if (sliceModelDisplayNode)
              {
              if (sliceModelDisplayNode->GetVisibility()
                && sliceModelDisplayNode->GetVisibility3D()
                && sliceModelDisplayNode->IsDisplayableInView(viewNode->GetID())
                && this->IsVolumeVisibleInSliceView(volumeNode, sliceLogic->GetSliceCompositeNode()))
                {
                // found a slice view that is displayed in this 3D view and the volume is visible on that slice
                visible = true;
                break;
                }
              }
```

For this to work, we would also need to observe display visibility changes of slice model nodes. Right now, when the eye icon is clicked in the slice view controller, this `UpdateActor` method is not called, so the color legend does not get displayed or hidden correctly (only when the volume node changes). Maybe you could give it a try to add these observers, but if it requires a lot of code or lots of additional observations then we can drop this and just keep the current simpler implementation (act if slice view was displayed in the 3D view).

Reply:
I will add observers to model nodes shortly.
If the eye icon is clicked and the slice view is showed up on a 3D view, the one can toggle visibility (Show Off / Show On) of color bar to update it visibility on a 3D.